### ISSUE TITLE
Looks for dist folder in wrong spot

### ISSUE THREAD

**Issue by spentak** [duffman/tspath#1]
Great work on this well needed package!

I think there may be an issue with resolving paths. With the correct setup for paths I get an error.

So I have:

```
Project
    src
    dist
```

My tsconfig:

```
        "outDir": "dist",
        "baseUrl": "./",
        "paths": {
            "api/*": ["src/api/*"],
            "utilities/*": ["src/utilities/*"],
            "middleware/*": ["src/middleware/*"],
            "resolvers/*": ["src/resolvers/*"],
            "graphql-schemas/*": ["src/graphql-schemas/*"]
        },
```
The line of code:

`import { graphqlSchema } from 'graphql-schemas/schemas';`

Trace:

`Error: Cannot find module '../src/graphql-schemas/schemas'`

Expected parsed path: `../graphql-schemas/schemas`

The paths that the parser are putting out are using the `baseUrl` as the root (it appears), but in reality they should be in relation to the directory in which they live? Like a reverse traverse upwards until they find the file?

**Comment by spentak** [comment]
Possible solution would be to allow an argument to the CLI --jsPath

`tspath -f --jsPath ./dist`  (instead of reading from tsconfig baseUrl)

**Comment by duffman** [comment]
HI there! I will reproduce your setup and look for possible solutions! brb :) thanks for helping me squash a potential bug!

**Comment by erikjalevik** [comment]
Unfortunately @spentak's solution doesn't work for me. I have:

```
"outDir": "dist",
"baseUrl": "./src"
```

and after building the project, I run:

```
./node_modules/.bin/tspath -f -jsPath ./dist
```

TSPath then tries to find the dist folder under src:

```
Error: ENOENT: no such file or directory, scandir: 'D:\...\projectName\src\dist'
```

It should be looking in 'D:\...\projectName\dist'.

**Comment by ShepelievD** [comment]
@erikjalevik the same

**Comment by ShepelievD** [comment]
@erikjalevik I have a decision, but my webstorm breaks, will answer you once success

**Comment by ShepelievD** [comment]
@erikjalevik what I did:

my `tsconfig.json`
```
{
  "compilerOptions": {
    "module": "commonjs",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "target": "esnext",
    "noImplicitAny": true,
    "moduleResolution": "node",
    "sourceMap": true,
    "outDir": "dist",
    "baseUrl": ".",
    "paths": {
      "@config/*": ["src/config/*"],
      "@models/*": ["src/models/*"],
      "@routes/*": ["src/routes/*"],
      "@src/types/*": ["src/types/*"],
      "@utils/*": ["src/utils/*"],
      "@graphql/*": ["src/graphql/*"],
      "*": [
        "node_modules/*"
      ]
    }
  },
  "include": [
    "src/**/*"
  ]
}
```

I made `tsconfig.path.json`

```
{
  "compilerOptions": {
    "outDir": "dist",
    "baseUrl": ".",
    "paths": {
      "@config/*": ["config/*"],
      "@models/*": ["models/*"],
      "@routes/*": ["routes/*"],
      "@src/types/*": ["types/*"],
      "@utils/*": ["utils/*"],
      "@graphql/*": ["graphql/*"],
      "*": ["node_modules/*"]
    }
  }
}
```

what I have to do is change `TS_CONFIG` to my custom own, ideally, we have to pass option to config file. But still it looks not quite good, I guess for first time we have to do it.

**Comment by ShepelievD** [comment]
@erikjalevik https://github.com/ShepelievD/tspath check it, example usage `tspath -f ./dist -c tsconfig.path.json`

Yeah, I know that is tricky, but I need it urgent :)

**Comment by kaanoo2904** [comment]
@ShepelievD 404 not found :(

**Comment by ShepelievD** [comment]
@kaanoo2904 yes, removed it, because decided to use https://github.com/ilearnio/module-alias

 `package.json`:
```
{
  ...
  "_moduleAliases": {
    "@config": "dist/config",
    "@models": "dist/models",
    "@routes": "dist/routes",
    "@types": "dist/types",
    "@utils": "dist/utils",
    "@graphql": "dist/graphql"
  },
  ...
}
```

`tsconfig.json`:
```
{
  "compilerOptions": {
    "module": "commonjs",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "target": "esnext",
    "noImplicitAny": true,
    "moduleResolution": "node",
    "sourceMap": true,
    "outDir": "dist",
    "baseUrl": "src",
    "paths": {
      "@config/*": ["config/*"],
      "@models/*": ["models/*"],
      "@routes/*": ["routes/*"],
      "@types/*": ["types/*"],
      "@utils/*": ["utils/*"],
      "@graphql/*": ["graphql/*"]
    }
  },
  "include": ["src/**/*"]
}
```

**Comment by iSuslov** [comment]
so.. it is useless because of this

**Comment by duffman** [comment]
> so.. it is useless because of this

No, you are pretty useless at reading and interpreting stuff I´d say... not even tspath can help you there

**Comment by duffman** [comment]
This has been resolved in 1.3.5

Regarding the suggestion of having tspath accept a param specifying the dist folder, I really don´t see when you´d use that? Being able to make it independent of the tsconfig.json would perhaps be something you could use in a headless build process, I don´t know but to single out the dist directory makes no real sense?

**Comment by 0x80** [comment]
@duffman 	I'm a little confused. I am experiencing the same issue as the OP using version 1.3.7. And if this has been resolved why is this issue still open?

My compiler options contain:

```json
{
    "outDir": "build",
    "baseUrl": ".",
    "paths": {
      "@src/*": ["src/*"],
      "@modules/*": ["src/modules/*"],
      "@utils/*": ["src/utils/*"]
    }
}
```

Paths in the build directory are all one level off. They contain "src" when they shouldn't. What would be a fix for this?

**Comment by nilswxa** [comment]
The fact that this is still unresolved is completely bananas. 🍌

**Comment by nilswxa** [comment]
For runtime resolution of /src/ files to a `build` or `dist` directory, I can confirm that https://github.com/ilearnio/module-alias works beautifully. Make sure to programmatically register your @app (or whatever) module to `__dirname`, so that it ends up looking under your build directory when resolving your absolute paths.

If you prefer 'build time' rewriting with a compiler plugin, the only `ttypescript` plugin that I actually got to work in this `src` scenario was https://github.com/zerkalica/zerollup. It gets the job done and requires no runtime magic, but keep in mind that from now on you need to tell your IDE (VS Code / IntelliJ / ...) to compile with `ttsc` rather than `typescript`. IntelliJ in particular is not a fan of this, so I abandoned compile time rewriting in favor of runtime resolution with a module alias.

I hope this is of help to someone. 🙂

**Comment by nilswxa** [comment]
@duffman: I was hoping that `tspath` would recognize the use of `rootDir` in `tsconfig.json` and strip it from the paths.

**Comment by 0x80** [comment]
@duffman Sorry for not being clear. With the tsconfig I use, `src` will not be part of the paths in the `build` (or dist) output. 

```ts
{
    "outDir": "build",
    "baseUrl": ".",
    "paths": {
      "@src/*": ["src/*"],
      "@modules/*": ["src/modules/*"],
      "@utils/*": ["src/utils/*"]
    },
	"include": ["src/**/*"]
}
```

A file at `/src/modules/foo.ts` will compile to `/build/modules/foo.js` and `/src/bar.ts` to `/build/bar.js`. If `bar.ts` was referencing `foo.ts` via path alias `@modules/bar`, the output of tspath will currently be `./src/modules/bar.js`.

Alternatively I have tried this config:


```ts
{
    "outDir": "build",
    "baseUrl": "./src",
    "paths": {
      "@src/*": ["./*"],
      "@modules/*": ["modules/*"],
      "@utils/*": ["utils/*"]
    },
	"include": ["src/**/*"]
}
```

This effectively results in the same output file structure. If I then try to run tspath, I get an error because it tries to read from /build/src.

So it seems to me there is an incompatibility between `tspath` and `tsc`, in how the tsconfig is interpreted. It is not something I would like or know how to solve by running a shell script or adding a build tool like Gulp.

**Comment by duffman** [comment]
What??? No that did not help anyone,  what in the world are you babbling
about?
You claimed my software to be "bananas" this rambling was more like
"bananas"

Do you have any actual feature request, please describe it in a structured
manner,
it is very weird to have someone call your software useless and get this
letter of a
madman as a response when I ask what the problem is...

Den mån 4 feb. 2019 kl 06:08 skrev Nils Weldon <notifications@github.com>:

**Comment by duffman** [comment]
To be clear my response was a reply to this:
"For runtime resolution of /src/ files to a build or dist directory, I can
confirm that https://github.com/ilearnio/module-alias works beautifully.
Make sure to programmatically register your @app <https://github.com/app> (or
whatever) module to __dirname, so that it ends up looking under your build
directory when resolving your absolute paths.

If you prefer 'build time' rewriting with a compiler plugin, the only
ttypescript plugin that I actually got to work in this src scenario was
https://github.com/zerkalica/zerollup. It gets the job done and requires no
runtime magic, but keep in mind that from now on you need to tell your IDE
(VS Code / IntelliJ / ...) to compile with ttsc rather than typescript.
IntelliJ in particular is not a fan of this, so I abandoned compile time
rewriting in favor of runtime resolution with a module alias.

I hope this is of help to someone. "

Den mån 4 feb. 2019 kl 14:47 skrev Patrik Forsberg <patfor@gmail.com>:

**Comment by Acathur** [comment]
Hi guys, I have created a pull request to solve this problem. #27

**Comment by jdtzmn** [comment]
@duffman Can we get @Acathur's PR merged? Are we waiting on something?

**Comment by JosephKrusling** [comment]
Since this is still unchanged, I recommend migrating to https://github.com/joonhocho/tscpaths

**Comment by duffman** [comment]
I Will Review the PR

ons 1 apr. 2020 kl. 18:58 skrev Joseph Krusling <notifications@github.com>:
